File: LeetCode/twopoint/RotateArray.java
package twopoint;

public class RotateArray {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        int partition = k % n;
        reverse(nums, 0, n - 1);
        reverse(nums, 0, partition - 1);
        reverse(nums, partition, n - 1);

    }

    private void reverse(int[] nums, int left, int right) {
        while (left < right) {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        RotateArray solution = new RotateArray();
        int[] nums = {1, 2, 3, 4, 5, 6, 7};
        int k = 3;
        solution.rotate(nums, k);
        for (int num : nums) {
            System.out.print(num + " ");
        }
        // Output: 5 6 7 1 2 3 4
    }
}

File: LeetCode_Python/twopoint/RotateArray.py
class RotateArray:
    def rotate(self, nums: list[int], k: int) -> None:
        n = len(nums)
        k %= n

        def reverse(left: int, right: int) -> None:
            while left < right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1

        reverse(0, n - 1)
        reverse(0, k - 1)
        reverse(k, n - 1)


if __name__ == "__main__":
    solution = RotateArray()

    nums = [1, 2, 3, 4, 5, 6, 7]
    solution.rotate(nums, 3)
    print("Expected Output: [5,6,7,1,2,3,4]")
    print("Actual Output:", nums)

    nums = [-1, -100, 3, 99]
    solution.rotate(nums, 2)
    print("Expected Output: [3,99,-1,-100]")
    print("Actual Output:", nums)
